src/libs/typesystem/SpecElektra/Elektra/SpecElektra.hs

```diff
@@ -0,0 +1,169 @@
+              | Any
+              | forall a . TyLft LftType a BaseType
+              | Sub CheckRestriction
+              | Sum BaseType BaseType
+              | BaseType :-> BaseType
+              | Bot
+
+type Path  = String
+type Value = Maybe String
+type TopUnit = ('Top 'Bot)
+
+-- *** some funky type level operators that makes it look nicer and less parantheses ***
+
+type f $ a = f a
+infixr 2 $
+
+type t <=> u = Intersection t u
+infixl 3 <=>
+
+type n ?> m = n $ 'Check m
+infixr 3 ?>
+
+type n ?+ m = n ':?+ 'Check m
+infixr 3 ?+
+
+type n <+> m = SubExpansion n ('Check m)
+infixr 3 <+>
+
+--
+-- *** Basic data and type definitions ***
+--
+
+-- The GADT that defines our keys and their properties in our type system
+-- It is basically an alternative evaluation/representation approach compared to the function approach down below
+data Key (a :: BaseType) = Key Path Value deriving Show
+
+-- Our Specifications. We only allow functions specified on our keys to enforce our BaseTypes
+-- Haskell Types are not allowed as we want to stay inside our BaseType Kind
+-- But in order to allow the optional addition of being able to calculate default values via normal haskell functions
+-- we just carry a second representation along. While the second representation would leak out haskell types and values
+-- and thus does not match our type system specification, the first parameter does
+data Spec (k :: BaseType) (b :: *)
+  where
+    Lft :: Key a -> Spec a (Key a)
+    Lam :: (Spec a (Key a) -> Spec c d) -> Spec (a ':-> c) (Key a -> d)
+    App :: Spec (a ':-> c) (Key a -> d) -> Spec a (Key a) -> Spec c d
+
+-- This is not required for the type system, its an optional extension to be able to support haskell functions directly
+-- to work with key values
+reduce :: Spec a b -> b
+reduce (Lft k)   = k
+reduce (Lam fn)  = \(x :: Key y) -> reduce (fn (Lft x))
+reduce (App f x) = reduce f (reduce x)
+
+eval :: (b ~ Key c) => Spec a b -> Key (TypeFinalizer c)
+eval = (\(Key p v) -> Key p v) . reduce
+
+--
+-- *** Our type system rules expressed in first order logic ***
+--
+type family SubExpansion (s :: BaseType) c :: BaseType where
+  SubExpansion ('Top TopUnit) x = 'Sub x
+  SubExpansion ('Sub s)       x = 'Sub $ s ':?+ x
+  SubExpansion _              _ = 'Bot
+
+-- Clears out the top/inferring states and resolves outstanding errors
+-- This is used to cleanup logical properties which are only relevant to the type system implementation
+type family TypeFinalizer (a :: BaseType) :: BaseType where
+  TypeFinalizer ('Top TopUnit)    = 'Any
+  TypeFinalizer ('Top a)          = a
+  TypeFinalizer ('Sum 'Bot _)     = 'Bot
+  TypeFinalizer ('Sum _ 'Bot)     = 'Bot
+  TypeFinalizer ('Sum a b)        = 'Sum (TypeFinalizer a) (TypeFinalizer b)
+  TypeFinalizer ('TyLft _ _ 'Bot) = 'Bot
+  TypeFinalizer ('TyLft a b c)    = ('TyLft a b (TypeFinalizer c))
+  TypeFinalizer a                 = a
+
+-- meaning of the rules from top to bottom:
```

Comment:
Why not put the comments directly to the rule? A more "literate haskell" style would be nice :smile:

Reply:
that file is not really used anymore, it was basically how the previous version of the typechecker worked. I am considering removing it as the previous branch still lives on in my repository anyway.